mypy/subtypes.py

```diff
@@ -2056,9 +2056,10 @@ def infer_variance(info: TypeInfo, i: int) -> bool:
         tvar = info.defn.type_vars[i]
         self_type = fill_typevars(info)
         for member in all_non_object_members(info):
-            # __mypy-replace is an implementation detail of the dataclass plugin
-            if member in ("__init__", "__new__", "__mypy-replace"):
+            if member in ("__init__", "__new__"):
                 continue
+            if "mypy-" in member:
```

Comment:
```suggestion
            if member.startswith("__mypy-"):
```

Reply:
I would also move `'__mypy-'` prefix to a constant somewhere

Reply:
Yes, defining a constant or a general function for filtering `mypy-` is something we should definitely do (if the general approach of this PR seems acceptable).  However, the commit suggestion would not work with the current state of this PR because the `dataclasses` plugin now mangles `__mypy-replace` and `__mypy-post_init`.